### ISSUE TITLE
Using Postgres objects rather than Information_Schema (W)

### ISSUE THREAD

**Issue by afig** [DASSL/ClassDB#1]
`metaFunctions.sql` uses `pg_views`, `prepareClassDB.sql` uses `pg_roles`. Using `Information_Schema` increases the use of standards compliant code (desirable), and makes it easier to port the scripts to another implementation.  

Evaluate the suitability of the applicable views provided by `Information_Schema` and implement them wherever possible.

**Comment by smurthys** [comment]
This issue is probably addressed, but a code review is required to make sure INFORMATION_SCHEMA is used everywhere possible and each use of pg_catalog is justified with suitable code comment.

**Comment by s-aithal** [comment]
I reviewed the code. All instances, except one, where `pg_catalog` is used are either for `pg_roles` or for `pg_has_role()`. The only exception is in Line 55 of `addUserMgmt.sql`, which uses `pg_catalog.pg_namespace`. I believe this can be replaced with `INFORMATION_SCHEMA.schemata`

**Comment by afig** [comment]
Thanks for the review @s-aithal. I agree that that the single instance of `pg_catalog.pg_namespace` could likely be replaced with `INFORMATION_SCHEMA.schemata`. As for replacing `pg_roles` and `pg_has_role()`, it seems like there is not a perfect analog for this in The Information Schema available in Postgres. 

One possible info schema view to replace those could be [`INFORMATION_SCHEMA.applicable_roles`](https://www.postgresql.org/docs/9.6/static/infoschema-applicable-roles.html). However, as the view's name implies, it only displays roles that the current user is a member of. As such, to replace our usage of `pg_roles` and `pg_has_role()`, it must be run under the context of a superuser, or possibly the `classdb` role. In that case, it seems like a helper function could be implemented to replace most or all usages of `pg_roles` an `pg_has_role()`.

**Comment by smurthys** [comment]
Good analyses @s-aithal and @afig. 

Unrelated to the specific topic of this conversation, but looking at function `createUser` in `addUserMgmt.sql`, the user in question is not granted rights if the schema already exists and no notice/exception is raised. I feel at least a notice should be raised in this event. If we agree this is a problem, I will create an issue.

**Comment by smurthys** [comment]
Love to get thoughts on [my comment from Aug. 17th](https://github.com/DASSL/ClassDB/issues/1#issuecomment-323181734) (the one immediately prior to this comment).

**Comment by afig** [comment]
I agree that raising an notice/exception would be the best course of action. Given that other parts of ClassDB depend on there being a correspondence between user and schema names, I believe that an exception would be best, so that there doesn't exist the possibility of there being a user without a corresponding schema they have access to. At minimum, a warning should be raised.

It might be helpful to view the table of [PostgreSQL's message severity levels](https://www.postgresql.org/docs/current/static/runtime-config-logging.html#RUNTIME-CONFIG-SEVERITY-LEVELS), copied here:

| Severity | Usage | syslog | eventlog |
|----------------|------------------------------------------------------------------------------------------------------|---------|-------------|
| DEBUG1..DEBUG5 | Provides successively-more-detailed information for use by developers. | DEBUG | INFORMATION |
| INFO | Provides information implicitly requested by the user, e.g., output from VACUUM VERBOSE. | INFO | INFORMATION |
| NOTICE | Provides information that might be helpful to users, e.g., notice of truncation of long identifiers. | NOTICE | INFORMATION |
| WARNING | Provides warnings of likely problems, e.g., COMMIT outside a transaction block. | NOTICE | WARNING |
| ERROR | Reports an error that caused the current command to abort. | WARNING | ERROR |
| LOG | Reports information of interest to administrators, e.g., checkpoint activity. | INFO | INFORMATION |
| FATAL | Reports an error that caused the current session to abort. | ERR | ERROR |
| PANIC | Reports an error that caused all database sessions to abort. | CRIT | ERROR |

A message with a severity of `ERROR` would be raised if we decide to go with an exception.

**Comment by smurthys** [comment]
I have created a new Issue #136.

@afig: It will be helpful if you recreate your informative comment containing the table of error levels at the new issue.

**Comment by smurthys** [comment]
I think we need to refocus on the original issue and make sure it is addressed, So, I propose a review by one of us, a confirmation by another, and then close this issue if it is addressed.
